core/src/test/java/com/linecorp/armeria/client/endpoint/dns/DnsAddressEndpointGroupTest.java

```diff
@@ -13,161 +13,273 @@
-        endpointGroup.query();
-        await().untilAsserted(
-                () -> assertThat(endpointGroup.endpoints()).containsExactlyElementsOf(ENDPOINTS_PORT));
+    public void backoff() throws Exception {
+        try (TestDnsServer server = new TestDnsServer(ImmutableMap.of())) { // Respond nothing.
+            try (DnsAddressEndpointGroup group = new DnsAddressEndpointGroupBuilder("backoff.com")
+                    .serverAddresses(server.addr())
+                    .resolvedAddressTypes(ResolvedAddressTypes.IPV4_PREFERRED)
+                    .backoff(Backoff.fixed(500))
+                    .build()) {
+
+                await().untilAsserted(() -> assertThat(group.attemptsSoFar).isGreaterThan(2));
+                assertThat(group.endpoints()).isEmpty();
+
+                // Start to respond correctly.
+                server.setResponses(ImmutableMap.of(
+                        new DefaultDnsQuestion("backoff.com.", A),
+                        new DefaultDnsResponse(0).addRecord(ANSWER, newAddressRecord("backoff.com", "1.1.1.1")),
+                        new DefaultDnsQuestion("backoff.com.", AAAA),
+                        new DefaultDnsResponse(0).addRecord(ANSWER, newAddressRecord("backoff.com", "::1"))));
+                assertThat(group.awaitInitialEndpoints()).containsExactly(
+                        Endpoint.of("backoff.com").withIpAddr("1.1.1.1"),
+                        Endpoint.of("backoff.com").withIpAddr("::1"));
+            }
+        }
     }
 
     @Test
-    public void unexpectedError() throws Exception {
-        endpointGroup = new DnsAddressEndpointGroup(
-                "armeria.com",
-                0,
-                resolver,
-                EVENT_LOOP,
-                Duration.ofSeconds(1));
-        endpointGroup.query();
-        await().untilAsserted(
-                () -> assertThat(endpointGroup.endpoints()).containsExactlyElementsOf(ENDPOINTS_NO_PORT));
-        when(resolver.resolveAll("armeria.com")).thenReturn(
-                EVENT_LOOP.newFailedFuture(new AnticipatedException("failed")));
-        endpointGroup.query();
-        // Unexpected errors ignored, do a small sleep to give time for the future to resolve.
-        Thread.sleep(TimeUnit.SECONDS.toMillis(1));
-        assertThat(endpointGroup.endpoints()).containsExactlyElementsOf(ENDPOINTS_NO_PORT);
+    public void backoffWithPartialResponse() throws Exception {
+        try (TestDnsServer server = new TestDnsServer(ImmutableMap.of(
+                // Respond A record only.
+                new DefaultDnsQuestion("backoff.com.", A),
+                new DefaultDnsResponse(0).addRecord(ANSWER, newAddressRecord("backoff.com", "1.1.1.1"))
+        ))) {
+            try (DnsAddressEndpointGroup group = new DnsAddressEndpointGroupBuilder("backoff.com")
+                    .serverAddresses(server.addr())
+                    .resolvedAddressTypes(ResolvedAddressTypes.IPV4_PREFERRED)
```

Comment:
I could be mistaken, but I can't find a test that doesn't call this - from my reading such a case would result in an assertion `!questions.isEmpty()` to fail.

Reply:
Adde a new test case `platformDefault`. Could you review again?